AssertionError: modelinv[4, 4] == 1 when displaying volume

Displaying a volume plot sometimes produces the following error:

```
julia> using GLMakie

julia> x = 17.318435178213235 .* (-1, 1)
(-17.318435178213235, 17.318435178213235)

julia> display(volume(x, x, x, rand(10,10,10)))
ERROR: AssertionError: modelinv[4, 4] == 1
Stacktrace:
  [1] (::GLMakie.var"#334#338"{Dict{Symbol, Any}})(modelinv::StaticArraysCore.SMatrix{4, 4, Float32, 16}, planes::Vector{Plane3f}, space::Symbol)
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/drawing_primitives.jl:896
  [2] map(::GLMakie.var"#334#338"{Dict{Symbol, Any}}, ::Union{Scene, Plot}, ::Observable{StaticArraysCore.SMatrix{4, 4, Float32, 16}}, ::Observable{Any}, ::Vararg{Observable{Any}}; ignore_equal_values::Bool, priority::Int64)
    @ Makie ~/.julia/packages/Makie/8h0bl/src/scenes.jl:167
  [3] map(::GLMakie.var"#334#338"{Dict{Symbol, Any}}, ::Union{Scene, Plot}, ::Observable{StaticArraysCore.SMatrix{4, 4, Float32, 16}}, ::Observable{Any}, ::Vararg{Observable{Any}})
    @ Makie ~/.julia/packages/Makie/8h0bl/src/scenes.jl:164
  [4] (::GLMakie.var"#332#336"{GLMakie.Screen{GLFW.Window}, Volume{Tuple{MakieCore.EndPoints{Float64}, MakieCore.EndPoints{Float64}, MakieCore.EndPoints{Float64}, Array{Float32, 3}}}})(gl_attributes::Dict{Symbol, Any})
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/drawing_primitives.jl:890
  [5] (::GLMakie.var"#210#220"{GLMakie.var"#332#336"{GLMakie.Screen{GLFW.Window}, Volume{Tuple{…}}}, GLMakie.Screen{GLFW.Window}, Scene, Volume{Tuple{MakieCore.EndPoints{…}, MakieCore.EndPoints{…}, MakieCore.EndPoints{…}, Array{…}}}})()
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/drawing_primitives.jl:337
  [6] get!(default::GLMakie.var"#210#220"{GLMakie.var"#332#336"{GLMakie.Screen{…}, Volume{…}}, GLMakie.Screen{GLFW.Window}, Scene, Volume{Tuple{…}}}, h::Dict{UInt64, GLMakie.GLAbstraction.RenderObject}, key::UInt64)
    @ Base ./dict.jl:479
  [7] cached_robj!(robj_func::GLMakie.var"#332#336"{GLMakie.Screen{GLFW.Window}, Volume{Tuple{…}}}, screen::GLMakie.Screen{GLFW.Window}, scene::Scene, plot::Volume{Tuple{MakieCore.EndPoints{…}, MakieCore.EndPoints{…}, MakieCore.EndPoints{…}, Array{…}}})
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/drawing_primitives.jl:235
  [8] draw_atomic(screen::GLMakie.Screen{GLFW.Window}, scene::Scene, plot::Volume{Tuple{MakieCore.EndPoints{Float64}, MakieCore.EndPoints{Float64}, MakieCore.EndPoints{Float64}, Array{Float32, 3}}})
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/drawing_primitives.jl:869
  [9] insert!(screen::GLMakie.Screen{GLFW.Window}, scene::Scene, x::Plot)
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/drawing_primitives.jl:355
 [10] insertplots!(screen::GLMakie.Screen{GLFW.Window}, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/screen.jl:515
 [11] insertplots!(screen::GLMakie.Screen{GLFW.Window}, scene::Scene) (repeats 2 times)
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/screen.jl:518
 [12] display_scene!(screen::GLMakie.Screen{GLFW.Window}, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/screen.jl:442
 [13] GLMakie.Screen(scene::Scene, config::GLMakie.ScreenConfig; visible::Nothing, start_renderloop::Bool)
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/screen.jl:459
 [14] GLMakie.Screen(scene::Scene, config::GLMakie.ScreenConfig)
    @ GLMakie ~/.julia/packages/GLMakie/GK1PF/src/screen.jl:455
 [15] getscreen(::Module, ::Scene, ::Dict{Symbol, Any})
    @ Makie ~/.julia/packages/Makie/8h0bl/src/display.jl:437
 [16] display(figlike::Makie.FigureAxisPlot; backend::Module, inline::Bool, update::Bool, screen_config::@Kwargs{})
    @ Makie ~/.julia/packages/Makie/8h0bl/src/display.jl:165
 [17] display(figlike::Makie.FigureAxisPlot)
    @ Makie ~/.julia/packages/Makie/8h0bl/src/display.jl:130
 [18] top-level scope
    @ REPL[3]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

This only happens for some values of x. I guess this is a floating point precision problem, and e.g. modelinv[4,4] = 0.99999 != 1.

- [x] what version of Makie are you running? v0.21.9
- [x] can you reproduce the bug with a fresh environment? Yes
- [x] What platform + GPU are you on? Linux

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 ## [Unreleased]
 
 - Reduce updates for image/heatmap, improving performance [#4130](https://github.com/MakieOrg/Makie.jl/pull/4130).
+- Fix float precision based assertions error in GLMakie.volume [#4311](https://github.com/MakieOrg/Makie.jl/pull/4311)
 
 ## [0.21.7] - 2024-08-19
 

--- a/GLMakie/src/drawing_primitives.jl
+++ b/GLMakie/src/drawing_primitives.jl
@@ -967,7 +967,7 @@ function draw_atomic(screen::Screen, scene::Scene, plot::Voxels)
             # with just applying it to the plane origin and transpose(inv(modelinv))
             # to plane.normal
             modelinv = inv(model)
-            @assert modelinv[4, 4] == 1
+            @assert (length(planes) == 0) || isapprox(modelinv[4, 4], 1, atol = 1e-6) 
 
             output = Vector{Vec4f}(undef, 8)
             for i in 1:min(length(planes), 8)

--- a/GLMakie/src/drawing_primitives.jl
+++ b/GLMakie/src/drawing_primitives.jl
@@ -893,7 +893,7 @@ function draw_atomic(screen::Screen, scene::Scene, plot::Volume)
             # model/modelinv has no perspective projection so we should be fine
             # with just applying it to the plane origin and transpose(inv(modelinv))
             # to plane.normal
-            @assert modelinv[4, 4] == 1
+            @assert (length(planes) == 0) || isapprox(modelinv[4, 4], 1, atol = 1e-6) 
 
             output = Vector{Vec4f}(undef, 8)
             for i in 1:min(length(planes), 8)

